File: Backend/shop/shop-typescript-nestjs-npm-mysql/src/cart/controller/cart.controller.ts
import { Body, Controller, Delete, Get, Param, ParseIntPipe, Patch, Post } from "@nestjs/common";
import { ApiBearerAuth, ApiOperation, ApiTags } from "@nestjs/swagger";
import { ok } from "../../common/api-response";
import { CurrentUser } from "../../common/current-user.decorator";
import type { CurrentUserPayload } from "../../common/current-user.decorator";
import { CartService } from "../service/cart.service";

@ApiTags("cart")
@ApiBearerAuth()
@Controller("api/v1/cart-items")
export class CartController {
  constructor(private readonly cartService: CartService) {}

  @Get()
  @ApiOperation({ summary: "List cart items" })
  async list(@CurrentUser() user: CurrentUserPayload) {
    return ok(await this.cartService.list(user));
  }

  @Post()
  @ApiOperation({ summary: "Create cart item" })
  async create(@CurrentUser() user: CurrentUserPayload, @Body() body: Record<string, unknown>) {
    return ok(await this.cartService.create(user, body));
  }

  @Patch(":cartItemId")
  @ApiOperation({ summary: "Update cart item" })
  async update(@CurrentUser() user: CurrentUserPayload, @Param("cartItemId", ParseIntPipe) cartItemId: number, @Body() body: Record<string, unknown>) {
    return ok(await this.cartService.update(user, cartItemId, body));
  }

  @Delete(":cartItemId")
  @ApiOperation({ summary: "Delete cart item" })
  async remove(@CurrentUser() user: CurrentUserPayload, @Param("cartItemId", ParseIntPipe) cartItemId: number) {
    return ok(await this.cartService.remove(user, cartItemId));
  }

  @Delete()
  @ApiOperation({ summary: "Clear cart" })
  async clear(@CurrentUser() user: CurrentUserPayload) {
    return ok(await this.cartService.clear(user));
  }
}

File: Backend/shop/shop-java-springboot-maven-postgresql/src/main/java/com/projectbible/shop/cart/controller/CartController.java
package com.projectbible.shop.cart.controller;

import com.projectbible.shop.cart.dto.CartDtos.CartItemResponseDto;
import com.projectbible.shop.cart.dto.CartDtos.UpsertCartItemDto;
import com.projectbible.shop.cart.service.CartService;
import com.projectbible.shop.common.api.ApiResponse;
import com.projectbible.shop.common.api.MessageResponse;
import com.projectbible.shop.common.security.AuthContext;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.servlet.http.HttpServletRequest;
import java.util.List;
import org.springframework.web.bind.annotation.*;

@Tag(name = "cart")
@RestController
@RequestMapping("/api/v1/cart-items")
public class CartController {
    private final CartService service;

    public CartController(CartService service) {
        this.service = service;
    }

    @Operation(summary = "List cart items")
    @GetMapping
    public ApiResponse<List<CartItemResponseDto>> list(HttpServletRequest request) {
        return ApiResponse.success(service.list(AuthContext.requireUser(request)));
    }

    @Operation(summary = "Add cart item")
    @PostMapping
    public ApiResponse<CartItemResponseDto> create(HttpServletRequest request, @RequestBody UpsertCartItemDto body) {
        return ApiResponse.success(service.create(AuthContext.requireUser(request), body));
    }

    @Operation(summary = "Update cart item")
    @PatchMapping("/{cartItemId}")
    public ApiResponse<CartItemResponseDto> update(HttpServletRequest request, @PathVariable long cartItemId, @RequestBody UpsertCartItemDto body) {
        return ApiResponse.success(service.update(AuthContext.requireUser(request), cartItemId, body));
    }

    @Operation(summary = "Delete cart item")
    @DeleteMapping("/{cartItemId}")
    public ApiResponse<MessageResponse> remove(HttpServletRequest request, @PathVariable long cartItemId) {
        return ApiResponse.success(service.remove(AuthContext.requireUser(request), cartItemId));
    }

    @Operation(summary = "Clear cart")
    @DeleteMapping
    public ApiResponse<MessageResponse> clear(HttpServletRequest request) {
        return ApiResponse.success(service.clear(AuthContext.requireUser(request)));
    }
}
